cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsValidator.java

```diff
@@ -0,0 +1,275 @@
+ * operation subsequent of OP6. However note that even OP1 and OP5 both outputs [x], OP1 output is
+ * not considered as invalid, since its used as input by OP2 which happens before OP5.
+ */
+public class LineageOperationsValidator {
+  private final Map<String, StageInputOutput> stageInputOutputs;
+  private final Map<String, List<PipelineOperation>> stageOperations;
+
+  public LineageOperationsValidator(Map<String, StageInputOutput> stageInputOutputs,
+                                    Map<String, List<PipelineOperation>> stageOperations) {
+    this.stageInputOutputs = stageInputOutputs;
+    this.stageOperations = stageOperations;
+  }
+
+  /**
+   * Validate the operations recorded by all stages.
+   * @throws InvalidLineageException if any invalid field is found.
+   */
+  void validate() {
+    Map<String, InvalidFieldOperations> invalidFieldOperationsForStage = new HashMap<>();
+
+    for (Map.Entry<String, List<PipelineOperation>> stageOperationEntry : stageOperations.entrySet()) {
+      String stageName = stageOperationEntry.getKey();
+      StageOperationsValidator validator = new StageOperationsValidator(stageInputOutputs.get(stageName),
+                                                                        stageOperationEntry.getValue());
+      validator.validate();
+      if (validator.getInvalidInputs().isEmpty() && validator.getInvalidOutputs().isEmpty()) {
+        continue;
+      }
+
+      InvalidFieldOperations invalids = new InvalidFieldOperations(validator.getInvalidInputs(),
+                                                                   validator.getInvalidOutputs());
+      invalidFieldOperationsForStage.put(stageName, invalids);
+    }
+
+    if (invalidFieldOperationsForStage.isEmpty()) {
+      return;
+    }
+
+    throw new InvalidLineageException(invalidFieldOperationsForStage);
+  }
+
+  /**
+   * Validate the operations recorded by single stage.
+   */
+  private static class StageOperationsValidator {
+    private final StageInputOutput stageInputOutput;
+    private final Map<String, List<String>> invalidOutputs;
+    private final Map<String, List<String>> invalidInputs;
+    private final List<PipelineOperation> operations;
+
+    StageOperationsValidator(StageInputOutput stageInputOutput, List<PipelineOperation> operations) {
+      this.stageInputOutput = stageInputOutput;
+      this.operations = operations;
+      this.invalidInputs = new HashMap<>();
+      this.invalidOutputs = new HashMap<>();
+    }
+
+    /**
+     * Validate the inputs and outputs for a stage. Please check the {@link LineageOperationsValidator}
+     * documentation for rules to figure out invalid fields.
+     */
+    void validate() {
+      // Fields input to the stage are valid
+      Set<String> validinputsSoFar = new HashSet<>(stageInputOutput.getInputs());
+
+      // Map of field name to the list of operations generated that field.
```

Comment:
operations generated that field -> operations that generated that field

Reply:
fixed.